### ISSUE TITLE
[Question] Sum within gradients

### ISSUE THREAD

**Issue by aoguedao** [ASEM000/Physics-informed-neural-network-in-JAX#1]
Hi Mahmoud! 

Thank you very much for this repo, I have been working on PINNs but only with third-party packages (deepxde) and now I have been wanting to implement my own with JAX.

I want to to something similar to the system of ODE you already have but I have been wondering why you sum the tensor t inside the gradient in the ode residual loss. 

```python
def ODE_loss(t,y1,y2):

  y1_t=lambda t:jax.grad(lambda t:jnp.sum(y1(t)))(t)
  y2_t=lambda t:jax.grad(lambda t:jnp.sum(y2(t)))(t)

  return y1_t(t) - y1(t)  , y2_t(t) - y1(t) + y2(t)
```

I have seen other implementations where then don't do that, but I think they use jacobian instead of grad.

Thank you very much!

**Comment by ASEM000** [comment]
Can you give me a pointer to the other implementations you mentioned?

**Comment by aoguedao** [comment]
Hi! Sure, no problem, for example this one https://github.com/erfanhamdi/pinn-torch/blob/main/Harmonic_Oscilator/main.py, line 45.

```python
def derivative(model: PhysicsNN, x_f: torch.Tensor, order: int = 1) -> torch.Tensor:
    """
    This function calculates the derivative of the model at x_f
    """
    dy = model(x_f)
    for i in range(order):
        dy = torch.autograd.grad(
            dy, x_f, grad_outputs = torch.ones_like(x_f), create_graph=True, retain_graph=True
        )[0]
    return dy
```

I have been reading around and some people recommend to use a combination of grad/jacobian and vmap, for example in the package `DeepXDE` , https://github.com/lululxvi/deepxde/blob/master/deepxde/gradients/gradients_forward.py#L77.

I am just learning JAX, so to be honest I still don't fully understand it.

Thanks!

**Comment by ASEM000** [comment]
Hello, 

The original implementation of PINNs used tensorflow `tf.gradient`, from the  tf docs for `tf.gradient` 

> gradients() adds ops to the graph to output the derivatives of ys with respect to xs. It returns a list of Tensor of length len(xs) where each tensor is the sum(dy/dx) for y in ys and for x in xs.

This means `tf.gradient` handles vector valued output by summing the  jacobian rows. `jax.grad` does not follow `tf.gradient` and is strictly defined for scalar output functions. One way to mimic the behavior of tf is by adding the sum ourselves. This is not the only way of course, The following snippets shows equivalent ways to mimic `tf.gradient` in `jax`/`tf`/`torch`.

_jax_

```python
import jax

x0, x1, x2 = 1.0, 2.0, 3.0
X = jax.numpy.array([x0, x1, x2])

def f(x):
    """R^3->R^2"""
    # x = [1, 2, 3]
    x0, x1, x2 = x
    # f0 = 1**2 + 2**2 = 1 + 4 = 5
    f0 = x0**2 + x1**2
    # f1 = 2**2 + 3**2 = 4 + 9 = 13
    f1 = x1**2 + x2**2
    return jax.numpy.array([f0, f1])

def vgrad_v1(f):
    """using grad of sum"""
    # grad of sum (e.g. sum(f0+f1))
    # R^3 -> R
    # df/dx0 = d(f0 + f1)/dx0 = df0/dx0 + df1/dx0
    # df/dx1 = ... = df0/dx1 + df1/dx1
    # df/dx2 = ... = df0/dx2 + df1/dx2
    return jax.grad(lambda x: jax.numpy.sum(f(x)))
df_1 = jax.jit(vgrad_v1(f))
print(f"grad of sum\n", df_1(X))

def vgrad_v2(f):
    """using jax.vjp"""
    def wrapper(x):
        y, vjp_fn = jax.vjp(f, x)
        return vjp_fn(jax.numpy.ones(y.shape))[0]
    return wrapper
df_2 = jax.jit(vgrad_v2(f))
print("vector valued grad with jax.vjp\n", df_2(X))

def vgrad_v3(f):
    """Using jax.jacfwd/rev"""
    # jacobian using jac{fwd,rev}
    # jacobian := dfi/dxj
    # [df0/dx0 df0/dx1 df0/dx2]  => [2*x0 2*x1 0   ]   => [2*1 2*2 0  ]  => [2 4 0]
    # [df1/dx0 df1/dx1  df1/dx2] => [0    2*x1 2*x2]   => [0   2*2 2*3]  => [0 4 6]
    # [1, 1] @ [[2, 4, 0] , [0, 4, 6]]
    def wrapper(x):
        J = jax.jacrev(f)(x)  # R^(out_dim x in_dim)
        ones = jax.numpy.ones(len(J))
        return ones @ J
    return wrapper
df_3 = jax.jit(vgrad_v3(f))
print("vector valued grad with jax.jacfwd/rev\n", df_3(X))
```

_tf_

```python

import tensorflow as tf

x0, x1, x2 = 1.0, 2.0, 3.0
X = tf.constant([x0, x1, x2])


@tf.function
def f(x):
    """R^3->R^2"""
    # x = [1, 2, 3]
    x0, x1, x2 = tf.unstack(x)
    # f0 = 1**2 + 2**2 = 1 + 4 = 5
    f0 = x0**2 + x1**2
    # f1 = 2**2 + 3**2 = 4 + 9 = 13
    f1 = x1**2 + x2**2
    return tf.stack([f0, f1])


def vgrad_v4(f):
    def wrapper(x):
        with tf.GradientTape() as tape:
            tape.watch(x)
            y = f(x)
        return tape.gradient(y, x)

    return wrapper

print("TF gradient\n", vgrad_v4(f)(X))

```

_torch_

```python

def vgrad_v5(f):
    def wrapper(x):
        y = f(x)
        return torch.autograd.grad(y, x, grad_outputs=torch.ones_like(y))[0]

    return wrapper


import torch

x0, x1, x2 = 1.0, 2.0, 3.0
X = torch.tensor([x0, x1, x2], requires_grad=True)


def f(x):
    """R^3->R^2"""
    # x = [1, 2, 3]
    x0, x1, x2 = x
    # f0 = 1**2 + 2**2 = 1 + 4 = 5
    f0 = x0**2 + x1**2
    # f1 = 2**2 + 3**2 = 4 + 9 = 13
    f1 = x1**2 + x2**2
    return torch.stack([f0, f1])


df_5 = vgrad_v5(f)
df_5(X)
```


P.S. For the code snippet, IIUC `grad_output` should be `ones_like` shape of the output not of the inputs. Nevertheless the code works in that instance because the model defined in that repo is R^1 -> R^1. 

P.S. In jax, vjp is more performant than v@Jacrev. You can see the flops used by each method in the following snippet

```python
df_1.lower(X).compile().cost_analysis()[0]["flops"] # 10.0
df_2.lower(X).compile().cost_analysis()[0]["flops"] # 10.0
df_3.lower(X).compile().cost_analysis()[0]["flops"] # 45.0
```

Let me know if this helps.

**Comment by aoguedao** [comment]
Thank you very much! I really appreciate your explanation and examples.

I had no idea tf.gradient does that. I am still not sure if this is what I need for my implementation but again, I do appreciate it!

**Comment by ASEM000** [comment]
No problem, let me know if you have any questions.
I will close for now.
